// react-navigation-tabs/lib/commonjs/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "createBottomTabNavigator", {
  enumerable: true,
  get: function get() {
    return _createBottomTabNavigator.default;
  }
});
Object.defineProperty(exports, "createMaterialTopTabNavigator", {
  enumerable: true,
  get: function get() {
    return _createMaterialTopTabNavigator.default;
  }
});
Object.defineProperty(exports, "BottomTabBar", {
  enumerable: true,
  get: function get() {
    return _BottomTabBar.default;
  }
});
Object.defineProperty(exports, "MaterialTopTabBar", {
  enumerable: true,
  get: function get() {
    return _MaterialTopTabBar.default;
  }
});
Object.defineProperty(exports, "createTabNavigator", {
  enumerable: true,
  get: function get() {
    return _createTabNavigator.default;
  }
});
Object.defineProperty(exports, "BottomTabBarProps", {
  enumerable: true,
  get: function get() {
    return _types.BottomTabBarProps;
  }
});
Object.defineProperty(exports, "NavigationTabState", {
  enumerable: true,
  get: function get() {
    return _types.NavigationTabState;
  }
});
Object.defineProperty(exports, "NavigationTabProp", {
  enumerable: true,
  get: function get() {
    return _types.NavigationTabProp;
  }
});
Object.defineProperty(exports, "NavigationTabScreenProps", {
  enumerable: true,
  get: function get() {
    return _types.NavigationTabScreenProps;
  }
});
Object.defineProperty(exports, "NavigationBottomTabOptions", {
  enumerable: true,
  get: function get() {
    return _types.NavigationBottomTabOptions;
  }
});
Object.defineProperty(exports, "NavigationMaterialTabOptions", {
  enumerable: true,
  get: function get() {
    return _types.NavigationMaterialTabOptions;
  }
});
Object.defineProperty(exports, "NavigationBottomTabScreenComponent", {
  enumerable: true,
  get: function get() {
    return _types.NavigationBottomTabScreenComponent;
  }
});
Object.defineProperty(exports, "NavigationMaterialTabScreenComponent", {
  enumerable: true,
  get: function get() {
    return _types.NavigationMaterialTabScreenComponent;
  }
});

var _createBottomTabNavigator = _interopRequireDefault(require("./navigators/createBottomTabNavigator"));

var _createMaterialTopTabNavigator = _interopRequireDefault(require("./navigators/createMaterialTopTabNavigator"));

var _BottomTabBar = _interopRequireDefault(require("./views/BottomTabBar"));

var _MaterialTopTabBar = _interopRequireDefault(require("./views/MaterialTopTabBar"));

var _createTabNavigator = _interopRequireDefault(require("./utils/createTabNavigator"));

var _types = require("./types");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
